### DISCUSSION TITLE
Best way to add a custom span processor on top of auto-configuration

### DISCUSSION THREAD

**Discussion by codeforaliving** [open-telemetry/opentelemetry-java#4595] [Q&A]
I would like to rename a child span after performing some logic. I understand this can be accomplished with a custom SpanProcessor.  What is the preferred way to add a custom SpanProcessor while still utilizing auto-config? I have seen mixed solutions and some seem deprecated. Is using an AutoConfigurationCustomizerProvider implementation the preferred way?  Does this require an extension or can it be handled in application code?  I am using Spring Boot.  Thanks in advance!

**Comment by mateuszrzeszutek** [comment]
Hey @codeforaliving ,
The `AutoConfigurationCustomizerProvider` is the preferred way of registering a custom `SpanProcessor`.
Are you using the javaagent? If so, you'll have to build an extension jar. If you're autoinitializing the OpenTelemetry SDK directly in your application code, then it is enough to just put your custom `AutoConfigurationCustomizerProvider` in the application code (also keep in mind the `META-INF/services` file).

**Reply by mateuszrzeszutek** [reply]
Just make sure that your application contains a `META-INF/services/<full-spi-name>` file that references your custom implementation. For example: https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure/src/testFullConfig/resources/META-INF/services/io.opentelemetry.sdk.autoconfigure.spi.AutoConfigurationCustomizerProvider

For reference, you can read more about Java SPI [here](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/ServiceLoader.html).

**Reply by codeforaliving** [reply]
Thanks for the response.  We are currently running the AWS distro of Open Telemetry for Java and run our service from the command line, using the following parameter: `-javaagent:/path/aws-opentelemetry-agent.jar`.  We don't have any manual instrumentation, rather we are relying on auto-instrumentation, but would like to "add" an additional `SpanProcessor` in code.  Does that mean we need an extension or can we use the `AutoConfigurationCustomizerProvider`?

**Reply by codeforaliving** [reply]
What is the deal with the `SpanNameExtractor`?

**Reply by mateuszrzeszutek** [reply]
> Does that mean we need an extension or can we use the `AutoConfigurationCustomizerProvider`?

If you're using the javaagent you'll have to create an extension - you can use `AutoConfigurationCustomizerProvider` in the extension jar to add a custom `SpanProcessor`.

You can find a demo/example extension project [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension) - it even contains a custom `SpanProcessor` implementation, you can just copy parts of it.

**Reply by mateuszrzeszutek** [reply]
> What is the deal with the `SpanNameExtractor`?

What do you mean? Is it related to https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/6299?

**Reply by codeforaliving** [reply]
Yes.  I am relatively new to OpenTelemetry as we are migrating from DataDog.  As we were going through and validating traces and spans, we noticed that span metrics generated from child spans had a generic name of "HTTP POST" and not the actual endpoint (path).  DataDog parsed and displayed this information as expected (for our workflow).  Moreover, we want to redact client information from the url path.  So something like `/api/service/customer/1234` shows up as `/api/service/customer/_` or something similar.  I am looking for the preferred and most supported way to achieve this goal.

**Reply by mateuszrzeszutek** [reply]
I think that a custom `SpanProcessor` is the way to go here.

**Reply by mateuszrzeszutek** [reply]
Correctly redacting/sanitizing the URL path requires some knowledge about how your application works, what kind of endpoints it calls, what pieces are the variables etc - and for that reason we're not doing this universally in the instrumentation.

**Reply by codeforaliving** [reply]
Ok, thanks for the reply.  Is creating a custom `SpanProcessor` experimental?  I would have to create a separate extension JAR correct?  In terms of stability, is it better to implement a custom processor in the collector contrib library or take the custom SpanProcessor route with an extension?

**Reply by codeforaliving** [reply]
Is there a way to initialize Open Telemetry in the application using auto-config (otel yaml config, env and props) but still add a custom SpanProcessor without an extension/external JAR?  I ask because we are trying to accomplish this logic in a common library that is shared by multiple services and it would be great if we didn't need to add a separate JAR.

**Reply by mateuszrzeszutek** [reply]
> Ok, thanks for the reply. Is creating a custom `SpanProcessor` experimental? ... In terms of stability, is it better to implement a custom processor in the collector contrib library or take the custom SpanProcessor route with an extension?

The `AutoConfigurationCustomizerProvider` is stable, `SpanProcessor` has been stable for a really long time too; the extension functionality in the javaagent is also stable. 
I'm not sure what kind of stability guarantees the collector makes about its custom processor, but you don't need to worry about the javaagent extension mechanism.

> I would have to create a separate extension JAR correct? 

Yes, that is correct.

**Reply by shadow0wolf** [reply]
can we use environment variables to set specific processors , in instrumented application ?
